File: iOS/ai/Data/ChatHistoryManager.swift
import Foundation

// MARK: - Chat History Manager

actor ChatHistoryManager {
    static let shared = ChatHistoryManager()

    private var historyDir: URL?
    private(set) var version: Int64 = 0

    private func ensureDir() -> URL {
        if let dir = historyDir { return dir }
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("chat-history", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        historyDir = dir
        return dir
    }

    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = .prettyPrinted
        return e
    }()

    private let decoder = JSONDecoder()

    // MARK: - Save

    func saveSession(_ session: ChatSession) -> Bool {
        let file = ensureDir().appendingPathComponent("\(session.id).json")
        guard let data = try? encoder.encode(session) else { return false }
        do {
            try data.write(to: file, options: .atomic)
            version = Int64(Date().timeIntervalSince1970 * 1000)
            return true
        } catch {
            return false
        }
    }

    // MARK: - Load

    func loadSession(_ sessionId: String) -> ChatSession? {
        let file = ensureDir().appendingPathComponent("\(sessionId).json")
        guard let data = try? Data(contentsOf: file) else { return nil }
        return try? decoder.decode(ChatSession.self, from: data)
    }

    func getAllSessions() -> [ChatSession] {
        let dir = ensureDir()
        guard let files = try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil) else { return [] }
        return files
            .filter { $0.pathExtension == "json" }
            .compactMap { try? decoder.decode(ChatSession.self, from: Data(contentsOf: $0)) }
            .sorted { $0.updatedAt > $1.updatedAt }
    }

    func getSessionCount() -> Int {
        let dir = ensureDir()
        let files = try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil)
        return files?.filter { $0.pathExtension == "json" }.count ?? 0
    }

    // MARK: - Delete

    func deleteSession(_ sessionId: String) -> Bool {
        let file = ensureDir().appendingPathComponent("\(sessionId).json")
        do {
            try FileManager.default.removeItem(at: file)
            version = Int64(Date().timeIntervalSince1970 * 1000)
            return true
        } catch {
            return false
        }
    }

    func clearHistory() {
        let dir = ensureDir()
        guard let files = try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil) else { return }
        var deleted = false
        for file in files {
            try? FileManager.default.removeItem(at: file)
            deleted = true
        }
        if deleted {
            version = Int64(Date().timeIntervalSince1970 * 1000)
        }
    }
}

File: android/app/src/main/java/com/ai/data/ChatHistoryManager.kt
package com.ai.data

import android.content.Context
import com.google.gson.Gson
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import kotlin.concurrent.withLock

/**
 * Singleton class to manage AI chat history storage.
 * Stores chat sessions as JSON files in the app's internal storage under a "chat-history" directory.
 */
object ChatHistoryManager {
    private const val HISTORY_DIR = "chat-history"
    private var historyDir: File? = null
    private val gson: Gson = createAppGson(prettyPrint = true)
    private val lock = java.util.concurrent.locks.ReentrantLock()
    private val _historyVersion = MutableStateFlow(0L)
    val historyVersion: StateFlow<Long> = _historyVersion.asStateFlow()

    /**
     * Initialize the manager with the app context.
     * Must be called before using any other methods.
     */
    fun init(context: Context) {
        lock.withLock {
            if (historyDir != null) return
            historyDir = File(context.filesDir, HISTORY_DIR).also { dir ->
                if (!dir.exists()) {
                    dir.mkdirs()
                }
            }
        }
    }

    /**
     * Save a chat session to storage.
     * If the session already exists (same ID), it will be updated.
     */
    fun saveSession(session: ChatSession): Boolean {
        val dir = historyDir ?: return false
        return lock.withLock {
            if (!dir.exists()) {
                dir.mkdirs()
            }
            val file = File(dir, "${session.id}.json")
            try {
                val json = gson.toJson(session)
                file.writeText(json)
                notifyHistoryChanged()
                android.util.Log.d("ChatHistoryManager", "Saved chat session: ${session.id}")
                true
            } catch (e: Exception) {
                android.util.Log.e("ChatHistoryManager", "Failed to save chat session: ${e.message}", e)
                false
            }
        }
    }

    /**
     * Load a chat session by ID.
     */
    fun loadSession(sessionId: String): ChatSession? {
        val dir = historyDir ?: return null
        return lock.withLock {
            val file = File(dir, "$sessionId.json")
            if (!file.exists()) return null
            try {
                val json = file.readText()
                gson.fromJson(json, ChatSession::class.java)
            } catch (e: Exception) {
                android.util.Log.e("ChatHistoryManager", "Failed to load chat session: ${e.message}", e)
                null
            }
        }
    }

    /**
     * Get all saved chat sessions, sorted by updatedAt (newest first).
     */
    fun getAllSessions(): List<ChatSession> {
        val dir = historyDir ?: return emptyList()
        if (!dir.exists()) return emptyList()

        return dir.listFiles { file -> file.extension == "json" }
            ?.mapNotNull { file ->
                try {
                    val json = file.readText()
                    gson.fromJson(json, ChatSession::class.java)
                } catch (e: Exception) {
                    android.util.Log.e("ChatHistoryManager", "Failed to parse chat session: ${e.message}")
                    null
                }
            }
            ?.sortedByDescending { it.updatedAt }
            ?: emptyList()
    }

    /**
     * Delete a chat session by ID.
     */
    fun deleteSession(sessionId: String): Boolean {
        val dir = historyDir ?: return false
        val file = File(dir, "$sessionId.json")
        return try {
            file.delete().also { deleted ->
                if (deleted) {
                    notifyHistoryChanged()
                }
            }
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Clear all chat history.
     */
    fun clearHistory() {
        val dir = historyDir ?: return
        var changed = false
        dir.listFiles()?.forEach { file ->
            changed = file.delete() || changed
        }
        if (changed) {
            notifyHistoryChanged()
        }
    }

    /**
     * Get the count of saved sessions.
     */
    fun getSessionCount(): Int {
        val dir = historyDir ?: return 0
        if (!dir.exists()) return 0
        return dir.listFiles { file -> file.extension == "json" }?.size ?: 0
    }

    suspend fun getAllSessionsAsync(): List<ChatSession> = withContext(Dispatchers.IO) {
        getAllSessions()
    }

    suspend fun getSessionCountAsync(): Int = withContext(Dispatchers.IO) {
        getSessionCount()
    }

    private fun notifyHistoryChanged() {
        _historyVersion.value = System.currentTimeMillis()
    }
}
